utils/utils/src/iter.rs

```diff
@@ -93,6 +93,68 @@ where
 {
 }
 
+pub struct FilterDrainIter<'a, F, T> {
+    vec: &'a mut Vec<T>,
+    pos: usize,
+    pred: F,
+}
+
+impl<'a, F, T> Iterator for FilterDrainIter<'a, F, T>
+where
+    F: Fn(&T) -> bool,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
I think the implementation is broken, because using `swap_remove`, the swapped elements will escape the test because `self.pos` is already past them. The test does not catch this by coincidence. For example if you append a `10` to your test vector you will see that it is broken.

Use `remove` instead of `swap_remove` would be an easy (though inefficent) fix.